pyteal/ast/subroutine.py

```diff
@@ -76,37 +76,176 @@ def __init__(
+            )
+        if not SubroutineDefinition.check_if_ABI_annotation(outputPType):
+            raise TealInputError(
+                "ABI type annotation for subroutine argument output get unexpected: {}".format(
+                    outputPType
+                )
+            )
+
     @staticmethod
     def _validate_parameter_type(
-        user_defined_annotations: dict, parameter_name: str
-    ) -> Type[Union[Expr, ScratchVar]]:
+        user_defined_annotations: dict, parameter_name: str, enforceABI: bool
+    ) -> Union[Type[Union[Expr, ScratchVar]], abi.TypeSpec]:
         ptype = user_defined_annotations.get(parameter_name, None)
+
+        if enforceABI and not SubroutineDefinition.check_if_ABI_annotation(ptype):
+            raise TealInputError(
+                f"Function has parameter {parameter_name} of not ABI type: {ptype}"
+            )
+
         if ptype is None:
-            # Without a type annotation, `SubroutineDefinition` presumes an implicit `Expr` declaration rather than these alternatives:
+            # Without a type annotation, `SubroutineDefinition` presumes an implicit `Expr` declaration
+            # rather than these alternatives:
             # * Throw error requiring type annotation.
             # * Defer parameter type checks until arguments provided during invocation.
             #
             # * Rationale:
             #   * Provide an upfront, best-effort type check before invocation.
-            #   * Preserve backwards compatibility with TEAL programs written when `Expr` is the only supported annotation type.
+            #   * Preserve backwards compatibility with TEAL programs written
+            #     when `Expr` is the only supported annotation type.
             # * `invoke` type checks provided arguments against parameter types to catch mismatches.
             return Expr
         else:
-            if not isclass(ptype):
+            if not isclass(ptype) and not SubroutineDefinition.check_if_ABI_annotation(
+                ptype
+            ):
                 raise TealInputError(
                     "Function has parameter {} of declared type {} which is not a class".format(
                         parameter_name, ptype
                     )
                 )
 
-            if ptype not in (Expr, ScratchVar):
+            if ptype in (Expr, ScratchVar):
+                return ptype
+            elif SubroutineDefinition.check_if_ABI_annotation(ptype):
+                return abi.type_spec_from_annotation(ptype)
+            else:
                 raise TealInputError(
                     "Function has parameter {} of disallowed type {}. Only the types {} are allowed".format(
-                        parameter_name, ptype, (Expr, ScratchVar)
+                        parameter_name, ptype, (Expr, ScratchVar, "ABI")
                     )
                 )
 
-            return ptype
+    @staticmethod
+    def _validate_full_signature(
```

Comment:
I recommend renaming to something that pays homage to the outputs. For example:

```python
def _arg_types_and_by_refs(...)
```
or if you think it's important to emphasize validation
```python
def _arg_types_and_by_refs_with_validation(...)
```

Reply:
I think the former suggestion makes sense.

Essentially the reason I move the chunk of code outside `__init__` to a separate function is that: the complexity proliferate when we introduce `ABI` types in subroutine.

Then out of the purpose of retrieving all `arg_types` and `args_by_ref`, it fine to rename to the one you suggested.